File: src/main/java/ru/citeck/ecos/index/IndexValue.java
package ru.citeck.ecos.index;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.impl.source.xml.XmlTokenImpl;
import com.intellij.psi.util.PsiUtil;
import com.intellij.util.io.DataExternalizer;
import icons.Icons;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.utils.EcosVirtualFileUtils;

import javax.swing.*;
import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RequiredArgsConstructor
@EqualsAndHashCode
public class IndexValue {

    private static final Icon DEFAULT_ICON = Icons.CiteckLogo;
    private static final Map<String, Icon> ICONS = Map.of(
            "alfresco", Icons.AlfrescoLogo
    );
    @Getter
    private final String id;
    @Getter
    private final Integer offset;
    @Getter
    private final String file;
    private Map<String, String> properties;

    public String getProperty(String property) {
        return properties == null ? null : properties.get(property);
    }

    public IndexValue setProperty(String property, String value) {
        if (properties == null) {
            properties = new HashMap<>();
        }
        properties.put(property, value);
        return this;
    }

    public PsiElement getPsiElement(Project project) {
        VirtualFile virtualFile = EcosVirtualFileUtils.getFileByPath(file);
        if (virtualFile == null) {
            return null;
        }
        PsiFile psiFile = PsiManager.getInstance(project).findFile(virtualFile);
        if (psiFile == null) {
            return null;
        }
        if (offset == 0) {
            return psiFile;
        }
        PsiElement element = PsiUtil.getElementAtOffset(psiFile, offset);
        if (element instanceof XmlTokenImpl) {
            return element.getParent();
        }
        return element;
    }

    public Icon getIcon() {
        String icon = getProperty("icon");
        if (icon != null) {
            return ICONS.getOrDefault(icon, DEFAULT_ICON);
        }
        return DEFAULT_ICON;
    }

    public static class Externalizer implements DataExternalizer<List<IndexValue>> {

        @Override
        public void save(@NotNull DataOutput out, List<IndexValue> indexValues) throws IOException {
            out.writeInt(indexValues.size());
            for (IndexValue indexValue : indexValues) {
                out.writeUTF(indexValue.id);
                out.writeInt(indexValue.offset);
                out.writeUTF(indexValue.file);
                Map<String, String> properties = indexValue.properties;
                if (properties == null) {
                    out.writeInt(0);
                } else {
                    out.writeInt(properties.size());
                    for (String property : properties.keySet()) {
                        out.writeUTF(property);
                        out.writeUTF(properties.get(property));
                    }
                }
            }
        }

        @Override
        public List<IndexValue> read(@NotNull DataInput in) throws IOException {
            List<IndexValue> indexValues = new ArrayList<>();
            int size = in.readInt();
            for (int counter = 0; counter < size; counter++) {
                IndexValue indexValue = new IndexValue(
                        in.readUTF(),
                        in.readInt(),
                        in.readUTF()
                );
                int propSize = in.readInt();
                for (int i = 0; i < propSize; i++) {
                    indexValue.setProperty(in.readUTF(), in.readUTF());
                }
                indexValues.add(indexValue);
            }
            return indexValues;
        }

    }

}

File: src/main/kotlin/ru/citeck/idea/search/index/IndexValue.kt
package ru.citeck.idea.search.index

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.psi.impl.source.xml.XmlTokenImpl
import com.intellij.psi.util.PsiUtil
import com.intellij.util.io.DataExternalizer
import icons.Icons
import ru.citeck.idea.utils.CiteckVirtualFileUtils.getFileByPath
import java.io.DataInput
import java.io.DataOutput
import javax.swing.Icon

data class IndexValue(
    val id: String,
    val offset: Int,
    val file: String
) {

    companion object {
        private val DEFAULT_ICON = Icons.CiteckLogo
        private val ICONS: Map<String, Icon> = java.util.Map.of()
    }

    private var properties: MutableMap<String, String>? = null

    fun getProperty(property: String): String? {
        return properties?.get(property)
    }

    fun setProperty(property: String, value: String): IndexValue {
        val props = properties ?: run {
            val newProps = HashMap<String, String>()
            properties = newProps
            newProps
        }
        props[property] = value
        return this
    }

    fun getPsiElement(project: Project): PsiElement? {
        val virtualFile = getFileByPath(file) ?: return null
        val psiFile = PsiManager.getInstance(project).findFile(virtualFile) ?: return null
        if (offset == 0) {
            return psiFile
        }
        val element = PsiUtil.getElementAtOffset(psiFile, offset)
        if (element is XmlTokenImpl) {
            return element.getParent()
        }
        return element
    }

    fun getIcon(): Icon {
        val icon = getProperty("icon")
        if (icon != null) {
            return ICONS.getOrDefault(icon, DEFAULT_ICON)
        }
        return DEFAULT_ICON
    }

    class Externalizer : DataExternalizer<List<IndexValue>> {

        override fun save(out: DataOutput, indexValues: List<IndexValue>) {
            out.writeInt(indexValues.size)
            for (indexValue in indexValues) {
                out.writeUTF(indexValue.id)
                out.writeInt(indexValue.offset)
                out.writeUTF(indexValue.file)
                val properties: Map<String, String>? = indexValue.properties
                if (properties == null) {
                    out.writeInt(0)
                } else {
                    out.writeInt(properties.size)
                    properties.forEach { (k, v) ->
                        out.writeUTF(k)
                        out.writeUTF(v)
                    }
                }
            }
        }

        override fun read(input: DataInput): List<IndexValue> {
            val indexValues: MutableList<IndexValue> = ArrayList()
            val size = input.readInt()
            for (counter in 0..<size) {
                val indexValue = IndexValue(
                    input.readUTF(),
                    input.readInt(),
                    input.readUTF()
                )
                val propSize = input.readInt()
                for (i in 0..< propSize) {
                    indexValue.setProperty(input.readUTF(), input.readUTF())
                }
                indexValues.add(indexValue)
            }
            return indexValues
        }
    }
}
